### PR TITLE
chore(slider): Refactor thumb interaction handlers

### PR DESCRIPTION
**Pull request by patrickrodee**
Expose slider interaction functionality (`handleDown_`, `handleMove_`, `handleUp_`) as methods on the foundation. By exposing the methods, porting the library to other frameworks will be easier. Yay!

### CODE REVIEW

**Comment by patrickrodee** [comment]
Adding "do-not-merge" for now since this dependent on #1782

File: packages/mdc-slider/foundation.js [hunk 1]
```diff
@@ -30,24 +32,38 @@ const KEY_IDS = {
   PAGE_DOWN: 'PageDown',
 };
 
-// Events that can constitute the user releasing drag on a slider
+/** @enum {string} */
+const EVENT_MAP = {
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: I would suggest renaming this to something more descriptive like `MOVE_EVENT_MAP`.

File: packages/mdc-slider/foundation.js [hunk 2]
```diff
@@ -75,11 +91,16 @@ export default class MDCSliderFoundation extends MDCFoundation {
...
       setLastTrackMarkersStyleProperty: (/* propertyName: string, value: string */) => {},
       isRTL: () => /* boolean */ false,
-    };
+    });
   }
 
-  constructor(adapter = {}) {
+  /**
+   * Creates a new instance of MDCSliderFoundation
+   * @param {MDCSliderAdapter} adapter
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Add nullability qualifier (`!` or `?`) before `MDCSliderAdapter`

File: packages/mdc-slider/foundation.js [hunk 3]
```diff
@@ -244,47 +270,82 @@ export default class MDCSliderFoundation extends MDCFoundation {
     }
   }
 
-  createDownHandler_(moveEvt, getPageX = ({pageX}) => pageX) {
+  /**
+   * Called when the user starts interacting with the slider
+   * @param {!Event} evt
+   */
+  handleDown_(evt) {
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Add `@private` annotation (ditto for other methods below)

File: packages/mdc-slider/foundation.js [hunk 4]
```diff
@@ -297,6 +358,10 @@ export default class MDCSliderFoundation extends MDCFoundation {
     return min + pctComplete * (max - min);
   }
 
+  /**
+   * Handles keydown events
+   * @param {Event} evt
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Add nullability qualifier (ditto for `Event` params in other methods below)

File: packages/mdc-slider/index.js [hunk 5]
```diff
@@ -17,51 +17,75 @@
...
+ * @extends MDCComponent<!MDCSliderFoundation>
+ */
+class MDCSlider extends MDCComponent {
   static attachTo(root) {
     return new MDCSlider(root);
   }
 
+  constructor(...args) {
+    super(...args);
+    /** @type {Element} */
```
Comment (acdvorak) [inline packages/mdc-slider/index.js]: Add `?` nullability qualifier (ditto for other fields below)

File: packages/mdc-slider/foundation.js [hunk 6]
```diff
@@ -144,13 +146,13 @@ class MDCSliderFoundation extends MDCFoundation {
   }
 
   destroy() {
-    this.adapter_.deregisterInteractionHandler('mousedown', this.mousedownHandler_);
-    this.adapter_.deregisterInteractionHandler('pointerdown', this.pointerdownHandler_);
-    this.adapter_.deregisterInteractionHandler('touchstart', this.touchstartHandler_);
+    DOWN_EVENTS.forEach((evtName) => {
+      this.adapter_.deregisterInteractionHandler(evtName, this.interactionStartHandler_);
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Ditto

File: packages/mdc-slider/foundation.js [hunk 7]
```diff
@@ -126,13 +130,11 @@ class MDCSliderFoundation extends MDCFoundation {
   init() {
     this.isDiscrete_ = this.adapter_.hasClass(cssClasses.IS_DISCRETE);
     this.hasTrackMarker_ = this.adapter_.hasClass(cssClasses.HAS_TRACK_MARKER);
-    this.adapter_.registerInteractionHandler('mousedown', this.mousedownHandler_);
-    this.adapter_.registerInteractionHandler('pointerdown', this.pointerdownHandler_);
-    this.adapter_.registerInteractionHandler('touchstart', this.touchstartHandler_);
+    DOWN_EVENTS.forEach((evtName) => this.adapter_.registerInteractionHandler(evtName, this.interactionStartHandler_));
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Can we combine this with line 137 so we only loop over `DOWN_EVENTS` once?

File: packages/mdc-slider/foundation.js [hunk 8]
```diff
@@ -269,53 +271,77 @@ class MDCSliderFoundation extends MDCFoundation {
   }
 
   /**
-   * Creates a handler for mouse/touch/pointer down events
-   * @param {string} moveEvt
-   * @param {(function(!Event): number)=} getPageX
-   * @return {function(!Event): undefined}
+   * Called when the user starts interacting with the slider
+   * @private
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: `@private` normally goes last, after all other annotations

File: packages/mdc-slider/foundation.js [hunk 9]
```diff
@@ -269,53 +271,77 @@ class MDCSliderFoundation extends MDCFoundation {
...
-      this.handlingThumbTargetEvt_ = false;
+    this.adapter_.registerBodyInteractionHandler(MOVE_EVENT_MAP[evt.type], moveHandler);
+    UP_EVENTS.forEach((evtName) => this.adapter_.registerBodyInteractionHandler(evtName, upHandler));
+    this.setValueFromEvt_(evt);
+  }
 
-      this.setActive_(true);
+  /**
+   * Called when the user moves the slider
+   * @private
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Ditto

File: packages/mdc-slider/foundation.js [hunk 10]
```diff
@@ -269,53 +271,77 @@ class MDCSliderFoundation extends MDCFoundation {
...
+   */
+  handleUp_() {
+    this.setActive_(false);
+    this.adapter_.notifyChange();
+  }
 
-    return downHandler;
+  /**
+   * Returns the pageX of the event
+   * @private
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Ditto

File: packages/mdc-slider/foundation.js [hunk 11]
```diff
@@ -269,53 +271,77 @@ class MDCSliderFoundation extends MDCFoundation {
...
+  getPageX_(evt) {
+    if (evt.targetTouches && evt.targetTouches.length > 0) {
+      return evt.targetTouches[0].pageX;
+    }
+    return evt.pageX;
   }
 
   /**
    * Sets the slider value from an event
+   * @private
```
Comment (acdvorak) [inline packages/mdc-slider/foundation.js]: Ditto

File: packages/mdc-slider/foundation.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (patrickrodee) [reply packages/mdc-slider/foundation.js]: That's a good point for discussion. I made them public eventually, then thought that `handleDown_`'s reliance on events wouldn't make sense to be called manually. But, now that we're talking about it, having them as public foundation methods makes sense.

**Pull request merged**